### ISSUE TITLE
Getting better images from DTD using PGAN

### ISSUE THREAD

**Issue by franclin** [facebookresearch/pytorch_GAN_zoo#112]
I am trying to generate some images using the pre-trained DTD on PGAN but the results are not looking very good.
Here is the pseudo-code that I am using as well as the results. Any thought on why the resulting images are not looking good?
Thanks in advance

```python
import torch
use_gpu = True if torch.cuda.is_available() else False
model = torch.hub.load('facebookresearch/pytorch_GAN_zoo:hub',
                       'PGAN', model_name='DTD',
                       pretrained=True, useGPU=use_gpu)
num_images = 4
noise, _ = model.buildNoiseData(num_images)
with torch.no_grad():
    generated_images = model.test(noise)

# let's plot these images using torchvision and matplotlib
import matplotlib.pyplot as plt
import torchvision
grid = torchvision.utils.make_grid(generated_images.clamp(min=-1, max=1), scale_each=True, normalize=True)
plt.imshow(grid.permute(1, 2, 0).cpu().numpy())
plt.show()
```

<img width="550" alt="Screenshot 2020-04-17 at 13 43 31" src="https://user-images.githubusercontent.com/2051702/79570994-4f3aa980-80b2-11ea-9eea-79757c4e443a.png">

**Comment by Molugan** [comment]
Hi,
For the DTD dataset, not all class are equally well trained. I would advise you to use the buildNoiseDataWithConstraints method to generate images from a specific class (see https://github.com/facebookresearch/pytorch_GAN_zoo/blob/7275ecbf53a9db7e4bc38c4c5136c10c4950724b/models/gan_visualizer.py#L156).

To see the available labels, see the section "attribKeysOrder" in the json configuration of your model. For DTD they are:
"attribKeysOrder": {
        "Main": {
            "order": 0,
            "values": [
                "banded",
                "blotchy",
                "braided",
                "bubbly",
                "bumpy",
                "chequered",
                "cobwebbed",
                "cracked",
                "crosshatched",
                "crystalline",
                "dotted",
                "fibrous",
                "flecked",
                "freckled",
                "frilly",
                "gauzy",
                "grid",
                "grooved",
                "honeycombed",
                "interlaced",
                "knitted",
                "lacelike",
                "lined",
                "marbled",
                "matted",
                "meshed",
                "paisley",
                "perforated",
                "pitted",
                "pleated",
                "polka-dotted",
                "porous",
                "potholed",
                "scaly",
                "smeared",
                "spiralled",
                "sprinkled",
                "stained",
                "stratified",
                "striped",
                "studded",
                "swirly",
                "veined",
                "waffled",
                "woven",
                "wrinkled",
                "zigzagged"
            ]
        }
    }

(The class name in "Main" and the values are the one listed above)

**Comment by Molugan** [comment]
I didn't notice: you must set getAvG = True in buildNoiseData.

**Comment by franclin** [comment]
Hi @Molugan 
 Thanks for your response, I tried your advice and got the same results. Is there a way to know the classes that were properly trained in this model please?
 Here is my pseudocode
```python
input = model.buildNoiseDataWithConstraints(num_images, {"Main": "fibrous"})
outImg = model.test(input, getAvG=True)
import matplotlib.pyplot as plt
import torchvision
grid = torchvision.utils.make_grid(outImg.clamp(min=0, max=1), scale_each=True, normalize=True)
plt.imshow(grid.permute(1, 2, 0).cpu().numpy())
```
Here are the results:
<img width="548" alt="Screen Shot 2020-05-14 at 21 26 16" src="https://user-images.githubusercontent.com/2051702/81982339-94a6b400-9629-11ea-8825-f287b8555a53.png">

**Comment by Molugan** [comment]
If I remember well, chequered, spiralled and lined work correctly.

**Comment by franclin** [comment]
Thanks for your prompt reply @Molugan . I tried those classes but the quality of the resulting images is still not good. I was wondering if I am just not missing a step in my pseudo-code that I mentioned earlier?
Here are the results fyi:
<img width="547" alt="Screen Shot 2020-05-16 at 10 07 28" src="https://user-images.githubusercontent.com/2051702/82115854-45fa3680-975d-11ea-9ff0-69f80dc68e73.png">

<img width="546" alt="Screen Shot 2020-05-16 at 10 06 03" src="https://user-images.githubusercontent.com/2051702/82115863-5e6a5100-975d-11ea-803b-6488805e00e9.png">

<img width="620" alt="Screen Shot 2020-05-16 at 10 04 49" src="https://user-images.githubusercontent.com/2051702/82115868-63c79b80-975d-11ea-9fa4-3e4ba7a5fa13.png">

**Comment by Molugan** [comment]
Hi,

I re-ran your code. 
The models of this library output values in the range [-1; 1], thus if you try to plot them with matplotlib without prior normalization you are likely to obtain ugly results. Here is an example of what you should be able to get with this model:
![image](https://user-images.githubusercontent.com/23697994/82191820-63610900-98f3-11ea-9eb9-ecf5171e137c.png)
![image](https://user-images.githubusercontent.com/23697994/82191900-7ecc1400-98f3-11ea-9b6a-c8dd39bb12d0.png)

To see an example of code to do the plot, you can have a look at https://github.com/facebookresearch/pytorch_GAN_zoo/blob/b75dee40918caabb4fe7ec561522717bf096a8cb/visualization/np_visualizer.py#L10

Can you try to launch this (ugly) snippet to see if it works ?
```
import numpy as np
import scipy
import scipy.misc
import imageio
import torch
from PIL import Image

def make_numpy_grid(arrays_list, gridMaxWidth=2048,
                    imgMinSize=128,
                    interpolation="nearest"):

    # NCWH format
    N, C, W, H = arrays_list.shape

    arrays_list = ((arrays_list + 1.0) * 255.0 / 2.0).astype(np.uint8)

    if C == 1:
        arrays_list = np.reshape(arrays_list, (N, W, H))

    gridMaxWidth = max(gridMaxWidth, W)

    imgSize = max(W, imgMinSize)
    imgHeight = int((float(imgSize) / W) * H)
    nImgsPerRows = min(N, int(gridMaxWidth // imgSize))

    gridWidth = nImgsPerRows * imgSize

    nRows = N // nImgsPerRows
    if N % nImgsPerRows > 0:
        nRows += 1

    gridHeight = nRows * imgHeight
    if C == 1:
        outGrid = np.zeros((gridHeight, gridWidth), dtype='uint8')
    else:
        outGrid = np.zeros((gridHeight, gridWidth, C), dtype='uint8')
    outGrid += 255

    interp = {
        'nearest': Image.NEAREST,
        'lanczos': Image.LANCZOS,
        'bilinear': Image.BILINEAR,
        'bicubic': Image.BICUBIC
    }

    indexImage = 0
    for r in range(nRows):
        for c in range(nImgsPerRows):

            if indexImage == N:
                break

            xStart = c * imgSize
            yStart = r * imgHeight

            img = np.array(arrays_list[indexImage])
            img = Image.fromarray(np.transpose(img, (1,2,0)))

            tmpImage = np.array(img.resize((imgSize, imgHeight), resample=interp[interpolation]))

            if C == 1:
                outGrid[yStart:(yStart + imgHeight),
                        xStart:(xStart + imgSize)] = tmpImage
            else:
                outGrid[yStart:(yStart + imgHeight),
                        xStart:(xStart + imgSize), :] = tmpImage

            indexImage += 1

    return outGrid

def saveImage(data, out_size_image, path, interpolation="bicubic"):

    if isinstance(out_size_image, tuple):
        out_size_image = out_size_image[0]
    data = torch.clamp(data, min=-1, max=1)
    outdata = make_numpy_grid(
        data.numpy(), imgMinSize=out_size_image, interpolation=interpolation)
    imageio.imwrite(path, outdata)

use_gpu = True if torch.cuda.is_available() else False
model = torch.hub.load("facebookresearch/pytorch_GAN_zoo:hub",
                       "PGAN", model_name="DTD",
                       pretrained=True, useGPU=use_gpu)
num_images = 8
classes = ["banded", "spiralled"]
for class_name in classes:
    noise = model.buildNoiseDataWithConstraints(num_images, {"Main":class_name})
    with torch.no_grad():
        generated_images = model.test(noise, getAvG=True)

    saveImage(generated_images, (128, 128), f"{class_name}.jpg")
```

**Comment by franclin** [comment]
Thanks for sharing this snippet, I ran it and it works fine now. Does it mean that the only classes that were properly trained were banded and spiralled? 
because this is what i got when trying chequered.

![chequered](https://user-images.githubusercontent.com/2051702/82711343-fea00880-9c7c-11ea-9fc0-1ae97450cad8.jpg)

Thanks for your help!

**Comment by Molugan** [comment]
Other classes works as well, for example striped will give that kind of results:
![image](https://user-images.githubusercontent.com/23697994/82712739-05cd1380-9c89-11ea-893f-cdd6228d6fed.png)

or cracked
![image](https://user-images.githubusercontent.com/23697994/82713051-f4383b80-9c89-11ea-8c11-0c04e1c16e90.png)

DTD is a really small dataset and thus the results can be very unequal. It's more like a demo GAN showing that you can get fancy results with few data (the swirly class for example is kind of funny).

**Comment by Molugan** [comment]
In other words, the best way to know if a class gives good enough results is just to try it out :)

**Comment by franclin** [comment]
thanks for your response, very happy.
